autonat/autonat-v2.md

```diff
@@ -0,0 +1,276 @@
+to the client node on the `/libp2p/autonat/2/dial` stream. The response contains
+a list of `DialStatus`es with a status for each address in the list up to and
+including the address that the server attempted to dial. The `DialStatus` for an
+address is set according to [Requirements for
+DialStatus](#requirements-for-dialstatus). The response also contains an
+appropriate `ResponseStatus` set according to [Requirements For
+ResponseStatus](#requirements-for-responsestatus).
+
+The client MUST check that the nonce received in the `DialAttempt` is the same
+as the nonce it sent in the `DialRequest`. If the nonce is different, it MUST
+discard this response.
+
+The server MUST close the stream after sending the response. The client MUST
+close the stream after receiving the response.
+
+
+### Requirements for DialStatus
+
+On receiving a `DialRequest` the server goes through the list of addresses in
+the request to select the first address that it is capable of dialing. For every
+address that the server checks, it assigns a `DialStatus` according to the
+following requirements. 
+
+For addresses that the server decides to not dial:
+
+`E_ADDRESS_UNKNOWN`: The server didn't understand the address. 
+
+`E_TRANSPORT_NOT_SUPPORTED`: The server understood the address, but has no
+transport capable of dialing the requested address. 
+
+`E_DIAL_REFUSED`: The server didn't dial the address because of address based
+restrictions like address based rate limit, the address being a private IP
+address, or a relay address.
+
+For the address that the server decided to dial:
+
+`E_DIAL_ERROR`: The server was unable to connect to the address
+
+`E_CONN_UPGRADE_FAILED`: The server was able to connect to the address, but
+failed to complete the connection upgrade step.
+
+`E_ATTEMPT_ERROR`: The server was able to establish an upgraded connection but
+some error occured when sending the nonce on the `/libp2p/autonat/2/attempt`
+stream.
+
+`OK`: The server successfully dialed the selected address.
+
+The server MUST NOT send dial statuses for addresses after the one it selected
+to dial.
+
+
+### Requirements for ResponseStatus
+
+The `ResponseStatus` sent by the server in the `DialResponse` message MUST be
+set according to the following requirements
+
+`OK`: the server completed the request successfully. A request is considered
+completed successfully when the server either completes a dial(successfully or
+unsuccessfully) or rejects all addresses in the request as undialable. 
+
+`E_BAD_REQUEST`: the server was unable to decode the request.
+
+`E_REQUEST_REFUSED`: the server didn't attempt to serve the request because of
+rate limiting, resource limit reached or blacklisting.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+server that prevented it from completing the request.
+
+Implementations MUST discard responses with status codes they do not understand. 
+
+### Amplification Attack Prevention
```

Comment:
@mxinden mention that this design is takes inspiration from a different protocol but I forgot which one. Might be worth mentioning!

Reply:
E.g. related to the QUIC anti-amplification mechanism based on data size:

> Prior to validating the client address, servers MUST NOT send more than three times as many bytes as the number of bytes they have received. This limits the magnitude of any amplification attack that can be mounted using spoofed source addresses.

https://www.rfc-editor.org/rfc/rfc9000.html#section-8.1

I am not aware of an exact origin of the AutoNATv2 mechanism.

Reply:
should I mention QUIC?